doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+  write with group commit).
+
+##### Hybrid Logical Timestamps
+
+Note: There might be an actual name for this concept, but I chose the name above because it’s similar to Hybrid Logical
+Clocks.
+
+Currently, the timestamps used to in Materialize for `upper`, `since`, and the global timestamp are 64 bit unsigned
+integers that are interpreted as an epoch in milliseconds. Instead we can change the representation of these timestamps
+so that the X most significant bits are used for the epoch and the Y least significant bits are used for a counter.
+Every 1 millisecond the X bits are increased by 1 and the Y bytes are all reset to 0. Every write increments the Y bits
+by 1. This would allow us to have 2^Y writes per millisecond, without blocking any queries. Any write in
+excess of 2^Y per millisecond must be blocked until the next millisecond.
+
+Note: Increasing the size of the timestamp type would give more flexibility (for example to 128 bits).
+
+**Pros:**
+
+- With a sufficiently high Y, no queries would be blocked.
+
+**Cons:**
+
+- Complicated to implement and likely has a large blast radius.
+- If X is low enough we will overflow our timestamps at some point in the future. This can be mitigated with a high
+  enough X.
+
+##### Write Ahead Log (WAL)
+
+This approach involves the coordinator buffering all writes in a durable write ahead log. All writes that occur in the
+same millisecond can be assigned the same timestamp, as long as there are no reads in-between them. Then at the end of
+the millisecond, or after X milliseconds, all writes are sent to STORAGE in a batch. If Materialize crashes/restarts
+before sending the writes to STORAGE, then it can just read the unsent write from the durable logs after starting back
+up. This prevents us from having to increment the global timestamp on every write. However, this does not solve the
+problem for read/write cycles.
+
+**Pros:**
+
+- No queries are blocked from consecutive writes.
+- A WAL would be useful for many other reasons not in scope for this document. It’s likely that no matter what solution
+  we pick, we’d still want to implement a WAL.
+
+**Cons:**
+
+- Only solves half the problem.
+
+##### Timely Fast Forward
+
+Timestamps in Materialize are meant to closely track the wall clock, but are not guaranteed to be exactly equal to the
+wall clock. This approach would involve forcibly fast-forwarding the `upper`s of all objects. Whenever a write would
+push the global timestamp above the wall clock by some threshold, all `upper`s are jumped to the global timestamp’s new
+value. The threshold could be set to 0 to prevent blocking any queries.
+
+**Pros:**
+
+- No queries are blocked by more than the assigned threshold.
+
+**Cons:**
+
+- Certain writes will require a broadcast to all nodes.
+- I’m not actually sure if this is possible, but if it is possible then it’s probably complicated and has a large blast
+  radius.
+
+## Alternatives
```

Comment:
Ideas I recall are maybe from Frank (or maybe not? I forget, or might be getting these details wrong):

- Today the default is that `since = upper - 1ms`. We can change stuff to hold back `since` to the lowest of all uppers (per timeline), or maybe hold it back to the global table timestamp. This would give us leeway on selecting read times maybe without forcibly making any query wait for things. Have to think about how it interacts with uppers?
- Only advance uppers (and/or sinces?) with the system clock. Unclear what to do for non realtime sources. Maybe they are ok because there's no tables involved so they can stay the same.